### PR TITLE
added socket support and in memory storage for online users

### PR DESCRIPTION
**Pull request by stevenkeezer**

### CODE REVIEW

File: client/src/pages/Chat.jsx [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
+import Users from '../components/Users';
+import { authenticationService } from '../services/authenticationService';
 
 export default function Chat() {
 	const history = useHistory();
 
-	React.useEffect(() => {}, [ history ]);
+	React.useEffect(
+		() => {
+			if (!authenticationService.currentUserValue) {
```
Comment (skassam21) [inline client/src/pages/Chat.jsx]: I believe we can't safely use currentUserValue from authenticationService as the current user information because the currentUser is loaded from localStorage. I explained this earlier in another PR, but the reason is because:
- the token may have expired and you are not checking this by just loading the currentUser from localStorage
- the user may have changed their information on another device and you'd need to update that on all localStorages of all devices

Instead, every time the application is first loaded, you need to make an API call to the back-end with the JWT token (which can be saved in localStorage though) to fetch the up to date user and to see if the token has expired. 

In order to do this, you may need to refactor some of your other work. I actually find having a user object stored in a context API makes life a bit easier

File: client/src/services/userService.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    const { enqueueSnackbar } = useSnackbar();
+    const handleResponse = useHandleResponse();
+    const requestOptions = {
+        method: 'GET',
+        headers: authHeader(),
+    };
+
+    const getUsers = () => {
+        return fetch(
+            `/api/users/userlist`,
```
Comment (skassam21) [inline client/src/services/userService.js]: Can we make this API call a bit more RESTful

File: server/controllers/userController.js [hunk 3]
```diff
@@ -1,8 +1,45 @@
...
 const validateRegisterInput = require("../validation/register")
 
+const onlineUsers = {};
+
+// @desc get list of users
+// @route POST /api/users/userlist
+// @acess Private
+const getUserList = asyncHandler(async (req, res) => {
+    try {
+        let jwtUser = jwt.verify(req.headers.authorization, process.env.JWT_SECRET);
```
Comment (skassam21) [inline server/controllers/userController.js]: This should happen in a middleware as you will be using this in a lot of functions

File: server/controllers/userController.js [hunk 4]
```diff
@@ -63,6 +105,14 @@ const registerUser = asyncHandler(async (req, res) => {
     }
 });
 
+// @desc   Logout a new user
+// @route   POST /api/users/logout
+// @access   Public
+const logout = asyncHandler(async (req, res) => {
+    const { currentUserId } = req.body
+    onlineUsers[currentUserId] = false
```
Comment (skassam21) [inline server/controllers/userController.js]: To avoid onlineUsers from becoming a really large object, you should delete a key from the onlineUsers object instead of setting it to false

File: server/controllers/userController.js [hunk 5]
```diff
@@ -1,8 +1,45 @@
...
+            })
+            .exec((err, users) => {
+                if (err) {
+                    console.log(err);
+                    res.setHeader("Content-Type", "application/json");
+                    res.end(JSON.stringify({ message: "Failure" }));
+                    res.sendStatus(500);
+                } else {
+                    res.send(users)
+                    req.io.sockets.emit("onlineUsers", onlineUsers);
```
Comment (skassam21) [inline server/controllers/userController.js]: Why do you need to emit this event here?

File: client/src/components/Users.jsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	);
+
+	useEffect(() => {
+		const socket = socketIOClient(process.env.REACT_APP_API_URL, {
+			transports: [ 'websocket', 'polling', 'flashsocket' ]
+		});
+		socket.on('onlineUsers', (data) => {
+			setOnlineUsers(data);
+		});
+		socket.on('users', (data) => {
```
Comment (skassam21) [inline client/src/components/Users.jsx]: I'd change this event name to be more readable about what it does (it is triggered when a new user is created right)?

File: client/src/components/Users.jsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+			setNewUser(data);
```
Comment (skassam21) [inline client/src/components/Users.jsx]: Instead of triggering a whole new fetch of all new users when a new user is created, you can just pass in the socket the user details, and then append that user to the users state variable

File: server/controllers/userController.js [hunk 8]
```diff
@@ -13,6 +50,8 @@ const authUser = asyncHandler(async (req, res) => {
     const user = await User.findOne({ email })
 
     if (user && (await user.matchPassword(password))) {
+        onlineUsers[user.id] = true
```
Comment (skassam21) [inline server/controllers/userController.js]: Technically a user is online when they have the application open, as opposed to when they are logged in. So instead, I would send events from the front-end React app when the user opens the application and when they close the application, to know if they are currently online

**Review by skassam21 — COMMENTED** [review]
This looks like a good starting point for online statuses! there were some structurally changes I'd like to see going forward. Removing the whole user object out of localStorage I would like to see going forward. In terms of tracking online users based on the application being open is something you can look at doing in the future.

File: server/controllers/userController.js (hunk 5 above)
Comment (stevenkeezer) [reply server/controllers/userController.js]: I was using this to update the online users whenever a user is logged in. I will be moving this away per your last comment, when a user connects & disconnects.

File: client/src/services/authenticationService.js [hunk 9]
```diff
@@ -3,18 +3,45 @@ import { useSnackbar } from 'notistack';
...
 };
 
+export function useVerify() {
+    const { enqueueSnackbar } = useSnackbar();
+
+    const verifyUser = (token) => {
+        const requestOptions = {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ token }),
```
Comment (skassam21) [inline client/src/services/authenticationService.js]: You should be using authHeader like you did for the getusers request

Comment (skassam21) [reply client/src/services/authenticationService.js]: Better practice to make the token in the header as a Bearer token

File: server/middleware/authMiddleware.js [hunk 10]
```diff
@@ -6,17 +6,16 @@ const protect = asyncHandler(async (req, res, next) => {
     let token;
 
     if (
-        req.headers.authorization &&
-        req.headers.authorization.startsWith("Bearer")
```
Comment (skassam21) [inline server/middleware/authMiddleware.js]: Why did you remove this? Instead maybe change the authHeader to include the word Bearer (it gives other developers an idea of the type of token being sent)

**Review by skassam21 — APPROVED** [review]
Just make a couple of these changes and then we are good to go forward!
